### DISCUSSION TITLE
How to import a TTF file as font which isn't installed on the system

### DISCUSSION THREAD

**Discussion by RVFET** [TomSchimansky/CustomTkinter#900] [Q&A] [answered]
I'm using a custom font and it isn't a system font, therefore it can't be imported by its name. Is there a way to import a font from an TTF or OTF file ?

**Comment by abhaymakes** [comment]
It *is* possible, however it is not integrated into tkinter or customtkinter, and is only possible by doing some hacks. Due to this, your font won't be platform independent.

Here is the sample code to use external fonts on windows:
```python
# Note : You'd still have to download the font in your system
import pyglet
import customtkinter

pyglet.font.add_file('file.ttf')

root = customtkinter.CTk()

MyLabel = customtkinter.CTkLabel(root,text="test",font=('font name',25))
MyLabel.pack()

root.mainloop()

# This is not going to work on linux
```

**Comment by sumeshir26** [comment]
There is a module called `extrafont` whixh does this on all systems.

**Comment by Akascape** [comment, marked as answer]
@RVFET No need to use any other package/module like pyglet, it's inbuilt in customtkinter.
```python
customtkinter.FontManager.load_font("my_font.ttf")
# then simply use that font
button = customtkinter.CTkButton(app, font=("my_font",20))
```
Make sure that the name of the font is properly described.

**Reply by ghost** [reply]
This only works on Windows and Linux and is accomplished by installing the font in the system. However, for macOS I believe the font the user passes ignored and SF Display is used, Apple's default. I got this from looking at `customtkinter/windows/widgets/font/font_manager.py`.

**Reply by BarakXYZ** [reply]
Is there a way to make it work on macOS as well?
I will test later this week if it actually doesn't work on my mac

**Reply by alejooroncoy** [reply]
And, did work on macOS? @BarakXYZ

**Reply by ghost** [reply]
I have modified `customtkinter/windows/widgets/font/font_manager.py` to allow for custom fonts to be loaded from files on Windows, macOS, and Linux. Here is the code:

```python
import sys
import os
import shutil
from typing import Union


class FontManager:

    linux_font_path = "~/.fonts/"

    @classmethod
    def init_font_manager(cls):
        # Linux
        if sys.platform.startswith("linux"):
            try:
                if not os.path.isdir(os.path.expanduser(cls.linux_font_path)):
                    os.mkdir(os.path.expanduser(cls.linux_font_path))
                return True
            except Exception as err:
                sys.stderr.write("FontManager error: " + str(err) + "\n")
                return False

        # other platforms
        else:
            return True

    @classmethod
    def windows_load_font(cls, font_path: Union[str, bytes], private: bool = True, enumerable: bool = False) -> bool:
        """ Function taken from: https://stackoverflow.com/questions/11993290/truly-custom-font-in-tkinter/30631309#30631309 """

        from ctypes import windll, byref, create_unicode_buffer, create_string_buffer

        FR_PRIVATE = 0x10
        FR_NOT_ENUM = 0x20

        if isinstance(font_path, bytes):
            path_buffer = create_string_buffer(font_path)
            add_font_resource_ex = windll.gdi32.AddFontResourceExA
        elif isinstance(font_path, str):
            path_buffer = create_unicode_buffer(font_path)
            add_font_resource_ex = windll.gdi32.AddFontResourceExW
        else:
            raise TypeError('font_path must be of type bytes or str')

        flags = (FR_PRIVATE if private else 0) | (FR_NOT_ENUM if not enumerable else 0)
        num_fonts_added = add_font_resource_ex(byref(path_buffer), flags, 0)
        return bool(min(num_fonts_added, 1))

    @classmethod
    def load_font(cls, font_path: str) -> bool:
        # Windows
        if sys.platform.startswith("win"):
            return cls.windows_load_font(font_path, private=True, enumerable=False)

        # Linux
        elif sys.platform.startswith("linux"):
            try:
                shutil.copy(font_path, os.path.expanduser(cls.linux_font_path))
                return True
            except Exception as err:
                sys.stderr.write("FontManager error: " + str(err) + "\n")
                return False

        # macOS
        elif sys.platform.startswith("darwin"):
            customtkinter_directory = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
            font_path = os.path.join(customtkinter_directory, "assets", "fonts", "noto_sans", "regular.ttf")
            try:
                # Define the path to the Fonts folder in the user's home directory
                fonts_folder = os.path.join(os.path.expanduser("~"), "Library", "Fonts")
                # Check if the Fonts folder exists, if not, create it
                if not os.path.exists(fonts_folder):
                    os.makedirs(fonts_folder)
                # Copy the font file to the Fonts folder
                shutil.copy(font_path, fonts_folder)
                return True
            except Exception as err:
                sys.stderr.write("FontManager error: " + str(err) + "\n")
                return False
            
        # others
        else:
            return
```
Not sure if it will be too helpful unless @TomSchimansky adds it.

**Reply by khalmurad** [reply]
> I have modified `customtkinter/windows/widgets/font/font_manager.py` to allow for custom fonts to be loaded from files on Windows, macOS, and Linux. Here is the code:
> 
> ```python
> import sys
> import os
> import shutil
> from typing import Union
> 
> 
> class FontManager:
> 
>     linux_font_path = "~/.fonts/"
> 
>     @classmethod
>     def init_font_manager(cls):
>         # Linux
>         if sys.platform.startswith("linux"):
>             try:
>                 if not os.path.isdir(os.path.expanduser(cls.linux_font_path)):
>                     os.mkdir(os.path.expanduser(cls.linux_font_path))
>                 return True
>             except Exception as err:
>                 sys.stderr.write("FontManager error: " + str(err) + "\n")
>                 return False
> 
>         # other platforms
>         else:
>             return True
> 
>     @classmethod
>     def windows_load_font(cls, font_path: Union[str, bytes], private: bool = True, enumerable: bool = False) -> bool:
>         """ Function taken from: https://stackoverflow.com/questions/11993290/truly-custom-font-in-tkinter/30631309#30631309 """
> 
>         from ctypes import windll, byref, create_unicode_buffer, create_string_buffer
> 
>         FR_PRIVATE = 0x10
>         FR_NOT_ENUM = 0x20
> 
>         if isinstance(font_path, bytes):
>             path_buffer = create_string_buffer(font_path)
>             add_font_resource_ex = windll.gdi32.AddFontResourceExA
>         elif isinstance(font_path, str):
>             path_buffer = create_unicode_buffer(font_path)
>             add_font_resource_ex = windll.gdi32.AddFontResourceExW
>         else:
>             raise TypeError('font_path must be of type bytes or str')
> 
>         flags = (FR_PRIVATE if private else 0) | (FR_NOT_ENUM if not enumerable else 0)
>         num_fonts_added = add_font_resource_ex(byref(path_buffer), flags, 0)
>         return bool(min(num_fonts_added, 1))
> 
>     @classmethod
>     def load_font(cls, font_path: str) -> bool:
>         # Windows
>         if sys.platform.startswith("win"):
>             return cls.windows_load_font(font_path, private=True, enumerable=False)
> 
>         # Linux
>         elif sys.platform.startswith("linux"):
>             try:
>                 shutil.copy(font_path, os.path.expanduser(cls.linux_font_path))
>                 return True
>             except Exception as err:
>                 sys.stderr.write("FontManager error: " + str(err) + "\n")
>                 return False
> 
>         # macOS
>         elif sys.platform.startswith("darwin"):
>             customtkinter_directory = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
>             font_path = os.path.join(customtkinter_directory, "assets", "fonts", "noto_sans", "regular.ttf")
>             try:
>                 # Define the path to the Fonts folder in the user's home directory
>                 fonts_folder = os.path.join(os.path.expanduser("~"), "Library", "Fonts")
>                 # Check if the Fonts folder exists, if not, create it
>                 if not os.path.exists(fonts_folder):
>                     os.makedirs(fonts_folder)
>                 # Copy the font file to the Fonts folder
>                 shutil.copy(font_path, fonts_folder)
>                 return True
>             except Exception as err:
>                 sys.stderr.write("FontManager error: " + str(err) + "\n")
>                 return False
>             
>         # others
>         else:
>             return
> ```
> 
> Not sure if it will be too helpful unless @TomSchimansky adds it.

Hi!
I also had a problem with the font on MacOS, and I had a solution like the one in the picture below:
![image](https://github.com/user-attachments/assets/5f198593-0bc6-464c-a978-c2145142d0c1)

But there is a problem. The first time I run my program it doesn't read custom fonts, the second time it does.
The custom font file is successfully copied on the first run of the program. I checked from the fonts folder, the font file is there. But the program doesn't show the custom font I need. The program reads the custom font without any problems on the next launches.

Can you recommend any solution to this problem?

**Reply by limafresh** [reply]
> @RVFET No need to use any other package/module like pyglet, it's inbuilt in customtkinter.
> 
> ```python
> customtkinter.FontManager.load_font("my_font.ttf")
> # then simply use that font
> button = customtkinter.CTkButton(app, font=("my_font",20))
> ```
> 
> Make sure that the name of the font is properly described.

Wow, thanks, Akascape, for this method! I've wanted to use fonts from files in my program for a long time, I searched the entire Internet (maybe I searched poorly) and didn't find how to do it. Some people wrote that Tkinter and CustomTkinter don't support it at all.

**Comment by VoidMaker06** [comment]
I would like to share with u guys that the font name and name of the font file is different, so if anyone tries to import the font file and it dosent load the font style, then try using the font name. For eg.  in the image below u can see the font name, i had imported the file but it was not loading the font so when i used the font name then only it loaded the font

![image](https://github.com/TomSchimansky/CustomTkinter/assets/137800983/9a14ac39-3671-434e-b3f6-e2b8e1a7b812)

**Reply by santvz6** [reply]
Muchas gracias :)

**Reply by stuffed18** [reply]
tysm that fixed my problem

**Comment by VoidMaker06** [comment]
Do you mean that on first run it only copies the font and in the second run
it loads it?

**Reply by khalmurad** [reply]
> Do you mean that on first run it only copies the font and in the second run it loads it?
> […](#)
> On Thu, Sep 12, 2024 at 10:58 AM Khalmurad ***@***.***> wrote: I have modified customtkinter/windows/widgets/font/font_manager.py to allow for custom fonts to be loaded from files on Windows, macOS, and Linux. Here is the code: import sysimport osimport shutilfrom typing import Union class FontManager: linux_font_path = "~/.fonts/" @classmethod def init_font_manager(cls): # Linux if sys.platform.startswith("linux"): try: if not os.path.isdir(os.path.expanduser(cls.linux_font_path)): os.mkdir(os.path.expanduser(cls.linux_font_path)) return True except Exception as err: sys.stderr.write("FontManager error: " + str(err) + "\n") return False # other platforms else: return True @classmethod def windows_load_font(cls, font_path: Union[str, bytes], private: bool = True, enumerable: bool = False) -> bool: """ Function taken from: https://stackoverflow.com/questions/11993290/truly-custom-font-in-tkinter/30631309#30631309 """ from ctypes import windll, byref, create_unicode_buffer, create_string_buffer FR_PRIVATE = 0x10 FR_NOT_ENUM = 0x20 if isinstance(font_path, bytes): path_buffer = create_string_buffer(font_path) add_font_resource_ex = windll.gdi32.AddFontResourceExA elif isinstance(font_path, str): path_buffer = create_unicode_buffer(font_path) add_font_resource_ex = windll.gdi32.AddFontResourceExW else: raise TypeError('font_path must be of type bytes or str') flags = (FR_PRIVATE if private else 0) | (FR_NOT_ENUM if not enumerable else 0) num_fonts_added = add_font_resource_ex(byref(path_buffer), flags, 0) return bool(min(num_fonts_added, 1)) @classmethod def load_font(cls, font_path: str) -> bool: # Windows if sys.platform.startswith("win"): return cls.windows_load_font(font_path, private=True, enumerable=False) # Linux elif sys.platform.startswith("linux"): try: shutil.copy(font_path, os.path.expanduser(cls.linux_font_path)) return True except Exception as err: sys.stderr.write("FontManager error: " + str(err) + "\n") return False # macOS elif sys.platform.startswith("darwin"): customtkinter_directory = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))) font_path = os.path.join(customtkinter_directory, "assets", "fonts", "noto_sans", "regular.ttf") try: # Define the path to the Fonts folder in the user's home directory fonts_folder = os.path.join(os.path.expanduser("~"), "Library", "Fonts") # Check if the Fonts folder exists, if not, create it if not os.path.exists(fonts_folder): os.makedirs(fonts_folder) # Copy the font file to the Fonts folder shutil.copy(font_path, fonts_folder) return True except Exception as err: sys.stderr.write("FontManager error: " + str(err) + "\n") return False # others else: return Not sure if it will be too helpful unless @TomSchimansky <https://github.com/TomSchimansky> adds it. Hi! I also had a problem with the font on MacOS, and I had a solution like the one in the picture below: image.png (view on web) <https://github.com/user-attachments/assets/5f198593-0bc6-464c-a978-c2145142d0c1> But there is a problem. The first time I run my program it doesn't read custom fonts, the second time it does. The custom font file is successfully copied on the first run of the program. I checked from the fonts folder, the font file is there. But the program doesn't show the custom font I need. The program reads the custom font without any problems on the next launches. Can you recommend any solution to this problem? — Reply to this email directly, view it on GitHub <[#900 (reply in thread)](https://github.com/TomSchimansky/CustomTkinter/discussions/900#discussioncomment-10621724)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/BA3K2F6VDKIOKUI23HGTP23ZWEQ7HAVCNFSM6AAAAAATDSVOA2VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANRSGE3TENA> . You are receiving this because you commented.Message ID: ***@***.*** .com>

Yes

Code:
![image](https://github.com/user-attachments/assets/d16925e8-61cc-4073-8524-0660f7493768)

**First run programm:**

Log result:
![image](https://github.com/user-attachments/assets/889ae54e-7a90-454d-9089-6f6fa9789d29)

There are fonts in ~/Library/Fonts folder (successfully copied):
<img width="636" alt="image" src="https://github.com/user-attachments/assets/feaccfa6-7b56-490b-895f-02314cc461b2">

Application also run:
<img width="604" alt="image" src="https://github.com/user-attachments/assets/060209f6-8cb0-4476-930e-67b34853c2a5">

But it is does not read my custom fonts


**Second run programm:**

Log result:
![image](https://github.com/user-attachments/assets/7ad8cf21-1b42-4f46-9eb7-0eba20b1bd2d)

Fonts are already copied:
<img width="627" alt="image" src="https://github.com/user-attachments/assets/74cb4fa4-a60d-4d39-9dae-5963b35bf784">

Application also run but with my custom fonts:
<img width="601" alt="image" src="https://github.com/user-attachments/assets/6c66c85b-622b-4419-838d-89fc63d173ad">

**Reply by khalmurad** [reply]
@VoidMaker06 Do you have any ideas?

**Reply by VoidMaker06** [reply]
What you could do is try checking if the fonts that are to be copied exists or not and after copying, if necessary, you can then load the fonts. You could try that

**Reply by VoidMaker06** [reply]
@khalmurad 

Edit: Lol, I didnt reply to you so i dont know if you got the ping from github about my reply so i added @ this time. If you did read my reply then did the idea work?

What you could do is try checking if the fonts that are to be copied exists or not and after copying, if necessary, you can then load the fonts. You could try that
